### ISSUE TITLE
OSX Keychain storage behavior in Keychain Access is strange

### ISSUE THREAD

**Issue by seanlutner** [docker/docker-credential-helpers#14]
I was testing out the new credential helpers this morning, specifically the osxkeychain method. I went through the example of logging in to hub.docker.com, and I saw that the Keychain Access entry was created. I also have an internal registry, which I was able to authenticate to and see an entry created, but only the username was stored. The login URL/host was not stored in the Where field, I only see http://, which is odd because my repo is SSL protected. See attached screenshots. I'm not sure that it matters operationally, but it would be nice if all registry authentications were stored properly. I might have the same username against multiple internal registries.

To login to the internal registry, I did `docker login my.registryhost.com:5000`

![Uploading registry1.png…]()
![Uploading registry2.png…]()

**Comment by calavera** [comment]
it looks like the screenshots were not attached, but I agree that it should work with any registry. I wonder if the URL is not properly populated in the docker side :pensive:

**Comment by seanlutner** [comment]
Whoops, I didn't realize they hadn't uploaded. Trying again. I have a related question, and I'll open another issue if required, does the linux helper require a full X environment based on it's dependency on gnome-keyring? I'm looking to use this from headless server systems and don't want to have to install full X everywhere.

![registry1](https://cloud.githubusercontent.com/assets/9679816/14560728/6d79b792-02df-11e6-9d85-f40e2b777bf6.png)
![registry2](https://cloud.githubusercontent.com/assets/9679816/14560730/7227687a-02df-11e6-8654-a0ad1ee21f92.png)

**Comment by dematic-denis-antonioli** [comment]
An other weird behavior I'm using the google container repository, logging in with
`docker login -u oauth2accesstoken -p "$(gcloud auth print-access-token)" us.gcr.io`
and the 'where' field is set to `http:///us.gcr.io`

I expect it to be `https://us.gcr.io`

<img width="259" alt="oauth2accesstoken" src="https://cloud.githubusercontent.com/assets/20247179/21191380/39548e78-c225-11e6-80c9-b51cd20bed63.png">

**Comment by KingEmet** [comment]
@dematic-denis-antonioli one workaround for that is to log in to 'https://us.gcr.io' explicitly:
`docker login -u oauth2accesstoken -p "$(gcloud auth print-access-token)" https://us.gcr.io`

**Comment by fajpunk** [comment]
aws-cli/1.11.61 Python/3.5.0 Darwin/15.6.0 botocore/1.5.24
Docker version 17.03.0-ce, build 60ccb22
osx 10.12.3

This is also affecting the awscli `aws ecr get-login` command on OSX.

Running this AWS CLI command gives this docker login command:

```
$ aws ecr get-login --profile myprofile --region us-east-1
docker login -u AWS -p mytoken -e none https://myaccountid.dkr.ecr.us-east-1.amazonaws.com
```

Running that docker login command puts an entry into the osx keychain with the
`Where` set to:

```
http:///myaccountid.dk.er.us-east-1.amazonaws.com
```

Note the incorrect protocol (`http` instead of `https`) and the three slashes
after the protocol, as previous comments have described.

Then, when I try to build an image from an image in that registry, with a
`FROM` line like this:

```
FROM myaccountid.dkr.ecr.us-east-1.amazonaws.com/appbase/python2.7.5
```

I get:

```
/usr/local/bin/docker build --rm -t myaccountid.dkr.ecr.us-east-1.amazonaws.com/myimage -f Dockerfile .
Sending build context to Docker daemon 1.045 MB
Step 1/11 : FROM myaccountid.dkr.ecr.us-east-1.amazonaws.com/myimage
Pulling repository myaccountid.dkr.ecr.us-east-1.amazonaws.com/myimage
unauthorized: authentication required
```

But if I manually edit the osx keychain entry and remove that third `/`, then
that pull succeeds.  So presumably, that third slash is making it so that the
keychain entry containing the login token can't be found.

**Comment by fajpunk** [comment]
One thing that could explain the reason for the wonky "Where" entry in the keychain is that docker-credential-osxkeychain expects the `ServerURL` value to be prefixed with a protocol (`http://` or `https://`), but docker passes just a domain, no protocol.

Because of this, `splitServer` (https://github.com/docker/docker-credential-helpers/blob/master/osxkeychain/osxkeychain_darwin.go#L137) is not parsing it correctly into the struct that gets passed to the OS command to add the entry to the keychain.

Specifically, `u.Scheme` will be blank here (https://github.com/docker/docker-credential-helpers/blob/master/osxkeychain/osxkeychain_darwin.go#L155), so will get defaulted to `http`, and because there's no protocol, url.Parse is putting the domain into `u.Path`, which I think explains the extra slash in the keychain entry.

**Comment by fajpunk** [comment]
For above ramblings:

`docker version`:
```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

`docker info`:

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 48
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.12-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: 4IRS:D5TV:GGV3:JM2Y:XVQ4:BOIS:6EC6:PJJI:HTR5:VR3P:RVAY:7W3G
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 27
 System Time: 2017-03-15T22:54:12.24670608Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by nfantis** [comment]
@fajpunk Thanks for the tip!  I was experiencing the same issue and removing the extra forward slash worked perfectly.

An alternative also worked, where I completely removed the following line from `.docker/config.json`:

`"credsStore": "osxkeychain"` (don't forget to delete the trailing comma on the line above)

Next, run:
`$(aws ecr get-login)`

The `.docker/config.json` file will be updated with the authentication token, bypassing Keychain all together.

**Comment by thaJeztah** [comment]
Fixed in version 0.5.2 through https://github.com/docker/docker-credential-helpers/issues/72
